language/move-lang/src/typing/globals.rs

```diff
@@ -265,27 +254,36 @@ where
     use TypeName_ as TN;
     use Type_ as T;
     let tloc = &global_type.loc;
-    let (def_loc, declared_module, sn, resource_opt) = match &global_type.value {
-        T::Var(_) => panic!("ICE type expansion failed"),
+    let (abilities, declared_module, sn) = match &global_type.value {
+        T::Var(_) | T::Apply(None, _, _) => panic!("ICE type expansion failed"),
         T::Anything | T::UnresolvedError => {
             return None;
         }
-        T::Apply(_, sp!(_, TN::ModuleType(m, s)), _args) => {
-            let def_loc = context.struct_declared_loc(m, s);
-            let resource_opt = context.resource_opt(m, s);
-            (def_loc, m.clone(), s, resource_opt)
+        T::Ref(_, _) | T::Unit => {
+            // Key ability is checked by constraints, and these types do not have Key
+            assert!(context.has_errors());
+            return None;
+        }
+        T::Apply(Some(abilities), sp!(_, TN::Multiple(_)), _)
+        | T::Apply(Some(abilities), sp!(_, TN::Builtin(_)), _) => {
+            // Key ability is checked by constraints
+            assert!(abilities.has_ability_(Ability_::Key).is_none());
+            assert!(context.has_errors());
+            return None;
         }
-        T::Ref(_, _)
-        | T::Unit
-        | T::Param(_)
-        | T::Apply(_, sp!(_, TN::Multiple(_)), _)
-        | T::Apply(_, sp!(_, TN::Builtin(_)), _) => {
+        T::Param(_) => {
             let ty_debug = core::error_format(global_type, &Subst::empty());
-            let tmsg = format!("Expected a nominal resource. Found the type: {}", ty_debug);
+            let tmsg = format!(
+                "Expected a struct type, global storage operations are restricted to struct types \
```

Comment:
Instead of a comma, this should have a period or maybe a colon. It sounds like a new sentence to me, though.